DeviceCommunications/DeviceComsProtocolParser.cs

```diff
@@ -0,0 +1,194 @@
+
+        public DeviceComsProtocolParser(IDeviceComsStream deviceComsStream)
+        {
+            _DeviceComsStream = deviceComsStream;
+
+            if (_DeviceComsStream.Connect())
+            {
+                // if device is connected then subscribe to it's DataReceived events
+                _DeviceComsStream.DataReceivedEvent += DataReceivedEventHandler;
+            }
+
+            _dataPoints = new ConcurrentQueue<(long, EEGData)>();
+        }
+
+        public event EventHandler DataUpdated;
+
+        void IDeviceComs.SendGetDataSinceLastRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x03;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        void IDeviceComs.SendStartDataCaptureRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x01;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        void IDeviceComs.SendStopDataCaptureRequest()
+        {
+            byte[] data = new byte[3];
+
+            data[0] = 0xFF;
+            data[1] = 0xFF;
+            data[2] = 0x02;
+
+            _DeviceComsStream.TransmitData(data);
+        }
+
+        private int _RxIndex = 0;
+        private byte[] _RxData = new byte[27];
+
+        EEGSignalMetadata IEEGData.SignalMetadata => throw new NotImplementedException();
+
+        long _LastUpdateTime;
+        long IEEGData.LastUpdateTime { get { return _LastUpdateTime; } }
+
+        /// <summary>
+        /// Also determines the interval at which this MockEEGData class will 
+        /// generate data.
+        /// Changing this has no effect if the timer is running.
+        /// </summary>
+        double IEEGData.DigitalMinimum { get; } = -1;
+
+        double IEEGData.DigitalMaximum { get; } = 1;
+
+        double IEEGData.PhysicalMinimum { get; } = -20;
+
+        double IEEGData.PhysicalMaximum { get; } = 20;
+
+        EEGSignalMetadata SignalMetadata { get; } = new EEGSignalMetadata()
+        {
+            Label = "EEG",
+            Units = "mV",
+            TransducerType = "UNKNOWN",
+            Prefiltering = "UNKNOWN",
+        };
+
+        // Returns a copy whenever it's accessed, to avoid concurrency issues.
+        public IEnumerable<(long, EEGData)> DataPoints
+        {
+            get
+            {
+                lock (_dataPointsLock)
+                {
+                    return _dataPoints.ToArray();
+                }
+            }
+        }
+
+        private void DataReceivedEventHandler(object sender, EventArgs e)
+        {
+            byte[] rxData;
+
+            int count = _DeviceComsStream.GetReceivedData(out rxData);
+
+            foreach (byte b in rxData)
+            {
+                _RxData[_RxIndex] = b;
+
+                switch (_RxIndex)
+                {
+                    case 0:
```

Comment:
It took me a while to understand this part, but I think it's to make sure we check for sync bytes correctly and process each packet when the whole thing is received while handling the fact that we will get 'random' chunks of data from the serial port (not necessarily a predictable 'packet at a time'), right? I may try to take a crack at seeing if this can be done more concisely, but if it works, it works!

Reply:
Actually now that I think about it, it's probably better to focus on your new protocol at the moment

Reply:
I've seen things like this implemented as a state machine which I like, this was only a quick way of getting it working.

Reply:
Gotcha! A state machine could be good